### PR TITLE
PCHR-2716: Create time amount picker directive

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

This PR add a custom Amount Time Picker (please do not confuse with a Timepicker Select directive). This directive is needed for Leave in Hours and may be reused in a lot of different places including Leave Request Modal, Job Contract leave tab etc.

## After

Note: this GIF was made in a temporary template for demonstration purposes only

## Technical Details

The directive can be used in the following way:

```js
<time-amount-picker
  time-amount-picker-value='vm.hours' // bind with a value
  time-amount-picker-min-amount='9.5' // optional, can be a variable, defaults to 0
  time-amount-picker-max-amount='18' // optional, can be a variable, defaults to 24
  time-amount-picker-interval='15' // optional, can be a variable, defaults to 1
></time-amount-picker>
```
`vm.hours` will be updated with an amount of hours (for example 11 or 15.75) once selectors changed.

----

✅ Jasmine Tests - passed
✅ JS distributive files - included
✅ Manual Tests - passed
⏹ Backstop Tests - omitted
⏹ PHP Unit Tests - n/a
⏹ CSS distributive files - n/a

### CODE REVIEW

File: org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+], function (_, directives) {
+  'use strict';
+
+  directives.directive('timeAmountPicker', ['$templateCache', function ($templateCache) {
+    return {
+      scope: {
+        timeAmountPickerMinAmount: '<',
+        timeAmountPickerMaxAmount: '<',
+        timeAmountPickerInterval: '<',
+        timeAmountPickerValue: '='
```
Comment (reneolivo) [inline org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js]: please use `minAmount`, `maxAmount`, `interval`, and `value`

File: org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      interval = +$scope.timeAmountPickerInterval || 1;
+      minAmount = +$scope.timeAmountPickerMinAmount || 0;
+      maxAmount = +$scope.timeAmountPickerMaxAmount || 24;
+
+      buildHoursOptions();
+      buildMinutesOptions();
+    }
+
+    /**
+     * Calculates the output humber in hours (float) and sets to the scope
```
Comment (reneolivo) [inline org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js]: humber -> number

File: org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+     */
+    function buildMinutesOptions () {
+      var minute = 0;
+
+      vm.minutesOptions = [];
+
+      while (minute < 60) {
+        var skip =
+          vm.selectedHours !== '' &&
+          ((+vm.selectedHours === Math.floor(minAmount) && minute < minAmount % 1 * 60) ||
```
Comment (reneolivo) [inline org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js]: I have no idea what's going on here tbh, can you please explain in the function description how this skips the value?

File: org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        return null;
+      }
+
+      $scope.timeAmountPickerValue = +vm.selectedHours + vm.selectedMinutes / 60;
+    }
+
+    /**
+     * Starts watching inbound attributes for changes
+     */
+    function startWatching () {
```
Comment (reneolivo) [inline org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js]: rename to `watchTimeAmountPickerOptions`

File: org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js (hunk 3 above)
Comment (reneolivo) [reply org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js]: Also wrapping `Math.floor(minAmount) && minute < minAmount % 1 * 60` into a variable would help.

File: org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+          vm.selectedHours !== '' &&
+          ((+vm.selectedHours === Math.floor(minAmount) && minute < minAmount % 1 * 60) ||
+           (+vm.selectedHours === Math.floor(maxAmount) && minute > maxAmount % 1 * 60));
+
+        (!skip) && vm.minutesOptions.push(minute);
+
+        minute += interval;
+      }
+
+      resetMinutesIfOutOfBounds();
```
Comment (reneolivo) [inline org.civicrm.reqangular/src/common/directives/time-amount-picker.directive.js]: can this reset be explained in the description too? thanks.

File: org.civicrm.reqangular/test/directives/time-amount-picker.directive.spec.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+      });
+    });
+
+    /**
+     * Builds a directive
+     *
+     * @param {Array} options objects of attributes keys, values and binding types
+     *   [{ key: 'key', value: 'value', bind: '<|@' }, {...}, ...]
+     */
+    function buildDirective (options) {
```
Comment (reneolivo) [inline org.civicrm.reqangular/test/directives/time-amount-picker.directive.spec.js]: In the future we might want to make a helper out of this `buildDirective` :)

Comment (igorpavlov) [reply org.civicrm.reqangular/test/directives/time-amount-picker.directive.spec.js]: Ah, you like it! 😍

**Pull request merged**